### ISSUE TITLE
Improve lua-ls typing for `TSConfig` class to stop enforcing optional fields as mandatory

### ISSUE THREAD

**Issue by schlerp** [nvim-treesitter/nvim-treesitter#5853]
**Is your feature request related to a problem? Please describe.**

TLDR; I believe the typing of `TSConfig` is incorrect.

I'm sick of the yucky diagnostics always shouting at me in my treesitter config. My config works fine without me setting those fields however the way they are typed in [configs.lua](https://github.com/nvim-treesitter/nvim-treesitter/blob/master/lua/nvim-treesitter/configs.lua#L11) is enforcing those fields as mandatory. I know the fields aren't mandatory because of the way we have coded the `setup()` function [here](https://github.com/nvim-treesitter/nvim-treesitter/blob/master/lua/nvim-treesitter/configs.lua#L410)

**Describe the solution you'd like**

I propose that we change from the current type definition of:

```lua
---@class TSConfig
---@field modules {[string]:TSModule}
---@field sync_install boolean
---@field ensure_installed string[]|string
---@field ignore_install string[]
---@field auto_install boolean
---@field parser_install_dir string|nil
```

To the following type definition of:

```lua
---@class TSConfig
---@field modules {[string]:TSModule}?
---@field sync_install boolean?
---@field ensure_installed string[]|string|nil
---@field ignore_install string[]?
---@field auto_install boolean?
---@field parser_install_dir string|nil
```

This will be compatible with all current code since it's allowing certain fields to be set to `nil` which is fine since we do a `nil` check and fall back to a default value via the `or` operator.

**Describe alternatives you've considered**

- I've considered just adding a comment to tell lua-ls to ignore the error for the following line, this would fix the issue for me only.
- I've considered just adding those fields into my config and setting them to empty values. But the language server will still complain since the typing will not accept `nil` as a valid value.

What we really want is for the type definition to be correct and support the valid runtime configuration of not having these values set; which for a simple lua table (eg. one without a metatable overwriting the default value returned for a missing key) is the same as setting them explicitly to `nil`.

I will submit a PR and reference this issue. I have made the change in my local copy of `nvim-treesitter` and can confirm it resolves the issue.

**Additional context**

before change:

<img width="1300" alt="image" src="https://github.com/nvim-treesitter/nvim-treesitter/assets/2208036/e79f0fe7-8db1-4b1a-965d-54045562ca67">

after change:

<img width="1300" alt="image" src="https://github.com/nvim-treesitter/nvim-treesitter/assets/2208036/35277dd6-4708-4315-98a2-efde099bf79b">

**Comment by clason** [comment]
The `master` branch is frozen, and these modules have been rewritten from scratch on the `main` branch (which will become nvim-treesitter 1.0). Any changes -- if they are still needed -- need to be made on the latter.

**Comment by schlerp** [comment]
Ahhh I see, not sure how I missed that! Skimming the doco i still can't see anything that indicates that.

It's a low priority change, just somethign thats beeng annoying me and i had some time to address. 

Can i ask, how long is it until v1.0 release?

**Comment by clason** [comment]
No ETA yet; rough goal is to make it the default after the Nvim 0.10 release (see https://github.com/nvim-treesitter/nvim-treesitter/issues/4767).

**Comment by schlerp** [comment]
ahh scratch that, i found it here: #4767

Wow, so I need to run nightly? I just dropped all the plugins that required that and swapped back to running latest stable. Hahah, ok no worries. I think for now i'll just use my fork until 1.0 comes out.

Thanks @clason for the update, sorry to waste your time!
